node-js-basics-101/node_modules/ws/lib/WebSocket.js

```diff
@@ -0,0 +1,712 @@
+
+  /**
+   * Set up the socket and the internal resources.
+   *
+   * @param {net.Socket} socket The network socket between the server and client
+   * @param {Buffer} head The first packet of the upgraded stream
+   * @private
+   */
+  setSocket (socket, head) {
+    socket.setTimeout(0);
+    socket.setNoDelay();
+
+    this._receiver = new Receiver(this.extensions, this.maxPayload, this.binaryType);
+    this._sender = new Sender(socket, this.extensions);
+    this._ultron = new Ultron(socket);
+    this._socket = socket;
+
+    // socket cleanup handlers
+    this._ultron.on('close', this._finalize);
+    this._ultron.on('error', this._finalize);
+    this._ultron.on('end', this._finalize);
+
+    // ensure that the head is added to the receiver
+    if (head && head.length > 0) {
+      socket.unshift(head);
+      head = null;
+    }
+
+    // subsequent packets are pushed to the receiver
+    this._ultron.on('data', (data) => {
+      this.bytesReceived += data.length;
+      this._receiver.add(data);
+    });
+
+    // receiver event handlers
+    this._receiver.onmessage = (data, flags) => this.emit('message', data, flags);
+    this._receiver.onping = (data, flags) => {
+      this.pong(data, !this._isServer, true);
+      this.emit('ping', data, flags);
+    };
+    this._receiver.onpong = (data, flags) => this.emit('pong', data, flags);
+    this._receiver.onclose = (code, reason) => {
+      this._closeMessage = reason;
+      this._closeCode = code;
+      this.close(code, reason);
+    };
+    this._receiver.onerror = (error, code) => {
+      // close the connection when the receiver reports a HyBi error code
+      this.close(code, '');
+      this.emit('error', error);
+    };
+
+    // sender event handlers
+    this._sender.onerror = (error) => {
+      this.close(1002, '');
+      this.emit('error', error);
+    };
+
+    this.readyState = WebSocket.OPEN;
+    this.emit('open');
+  }
+
+  /**
+   * Clean up and release internal resources.
+   *
+   * @param {(Boolean|Error)} Indicates whether or not an error occurred
+   * @private
+   */
+  finalize (error) {
+    if (this._finalizeCalled) return;
+
+    this.readyState = WebSocket.CLOSING;
+    this._finalizeCalled = true;
+
+    clearTimeout(this._closeTimer);
+    this._closeTimer = null;
+
+    //
+    // If the connection was closed abnormally (with an error), or if the close
+    // control frame was malformed or not received then the close code must be
+    // 1006.
+    //
+    if (error) this._closeCode = 1006;
+
+    if (this._socket) {
+      this._ultron.destroy();
+      this._socket.on('error', function onerror () {
+        this.destroy();
+      });
+
+      if (!error) this._socket.end();
+      else this._socket.destroy();
+
+      this._socket = null;
+      this._ultron = null;
+    }
+
+    if (this._sender) {
+      this._sender = this._sender.onerror = null;
+    }
+
+    if (this._receiver) {
+      this._receiver.cleanup(() => this.emitClose());
```

Comment:
__[jshint]__
* 'arrow function syntax (=>)' is only available in ES6 (use 'esversion: 6').

<!-- Comment by [SideCI](https://sideci.com) -->